codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoDependency.java

```diff
@@ -15,82 +15,233 @@
+        return version;
     }
 
     @Override
     public List<SymbolDependency> getDependencies() {
-        return Collections.singletonList(dependency);
+        Set<SymbolDependency> symbolDependencySet = new TreeSet<>(SetUtils.of(getSymbolDependency()));
+
+        symbolDependencySet.addAll(resolveDependencies(getGoDependencies(), new TreeSet<>(SetUtils.of(this))));
+
+        return new ArrayList<>(symbolDependencySet);
+    }
+
+    private Set<SymbolDependency> resolveDependencies(Set<GoDependency> goDependencies, Set<GoDependency> processed) {
+        Set<SymbolDependency> symbolDependencies = new TreeSet<>();
+        if (goDependencies.size() == 0) {
+            return symbolDependencies;
+        }
+
+        Set<GoDependency> dependenciesToResolve = new TreeSet<>();
+        for (GoDependency dependency : goDependencies) {
+            if (processed.contains(dependency)) {
+                continue;
+            }
+            processed.add(dependency);
+            symbolDependencies.add(dependency.getSymbolDependency());
+            dependenciesToResolve.addAll(dependency.getGoDependencies());
+        }
+
+        symbolDependencies.addAll(resolveDependencies(dependenciesToResolve, processed));
+
+        return symbolDependencies;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        } else if (!(o instanceof GoDependency)) {
+            return false;
+        }
+
+        GoDependency other = (GoDependency) o;
+
+        return this.type.equals(other.type) && this.sourcePath.equals(other.sourcePath)
+                && this.importPath.equals(other.importPath) && this.alias.equals(other.alias)
+                && this.version.equals(other.version) && this.dependencies.equals(other.dependencies);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(type, sourcePath, importPath, alias, version, dependencies);
+    }
+
+    @Override
+    public int compareTo(GoDependency o) {
+        if (equals(o)) {
+            return 0;
+        }
+
+        int importPathCompare = importPath.compareTo(o.importPath);
+        if (importPathCompare != 0) {
+            return importPathCompare;
+        }
+
+        if (alias != null) {
+            return alias.compareTo(o.alias);
+        }
+
+        return version.compareTo(o.version);
+    }
+
+    public enum Type {
+        STANDARD_LIBRARY, DEPENDENCY;
+
+        @Override
+        public String toString() {
+            switch (this) {
+                case STANDARD_LIBRARY:
+                    return "stdlib";
+                case DEPENDENCY:
+                    return "dependency";
+                default:
+                    return "unknown";
+            }
+        }
+    }
+
+    public static GoDependency moduleDependency(String sourcePath, String importPath, String version, String alias) {
```

Comment:
Minor nit, it would be handy for this function to be overridden with a form that doesn't take the `alias` parameter, or defaults `alias` to null.